Editorial:  Just Green Enough (#6330)
cpinitiative/usaco-guide: solutions/silver/usaco-1112.mdx

---
id: usaco-1112
source: USACO Silver 2021 February
title: Just Green Enough
author: John Zhou
---

[Official Analysis (C++, Java)](https://usaco.org/current/data/sol_prob3_silver_feb21.html)

> 

Reviewer: ```suggestion

## Explanation

```

Firstly, let's consider solving this problem in 1D, where we pick intervals from the list $g(j)$, the values for an $1\times N$ pasture.

Let $g'(j)$ be $-1$ if $g(j)<100$, $0$ if $g(j)=100$, and $1$ if $g(j)>100$.

> We must find a range of $j$'s in which there are no $-1$'s, and at least $1$ $0$.

Reviewer: ```suggestion
We must find a range of $j$'s in which there are no $-1$'s, and at least one $0$.
```

We precompute $g'(j)$.

> The objective is now counting the amount of intervals with no $-1$'s and at least $1$ $0$.

Reviewer: ```suggestion
The objective is now counting the amount of intervals with no $-1$'s and at least one $0$.
```

---
id: usaco-1112
source: USACO Silver 2021 February
title: Just Green Enough
author: John Zhou
---

[Official Analysis (C++, Java)](https://usaco.org/current/data/sol_prob3_silver_feb21.html)

## Explanation

Firstly, let's consider solving this problem in 1D, where we pick intervals from the list $g(j)$, the values for an $1\times N$ pasture.

Let $g'(j)$ be $-1$ if $g(j)<100$, $0$ if $g(j)=100$, and $1$ if $g(j)>100$.
We must find a range of $j$'s in which there are no $-1$'s, and at least one $0$.
We precompute $g'(j)$.
The objective is now counting the number of intervals with no $-1$'s and at least one $0$.

Next, we sweep $j$ from $1$ to $N$, and keep track of the last index we encounter $0$ and $-1$'s as $j_0$ and $j_{-1}$, respectively.
For an interval ending at $j$ to be picked, it must include a $0$, so the starting index must be no later than the last $0$ we encounter;
it also must not include a $-1$, so the starting index must be later than the last $-1$ we encounter.

> Thus, any starting index in the range $(j_{-1}, j_0]$ works, provided it is not degenerate, so we simply accumulate our current value of $\max(0, j_0-j_{-1})$ while we sweep.

Reviewer: many may not know what you mean by degenerate here. little wording would be nice.

Author: ```suggestion
Thus, any starting index in the range $(j_{-1}, j_0]$ works provided $j_{-1}<j_0$, so we simply accumulate our current value of $\max(0, j_0-j_{-1})$ while we sweep.
```

> We accumulate if and only if $g'(j)\neq -1$, as a value of $-1$ means we cannot end with this element.

Reviewer: "if and only if" is a bit confusing. it is worth mentioning that g`(j) belongs to {0, 1} along with it.

Author: ```suggestion
We accumulate at $j$ only if $g'(j)\in{0,1}$, as we cannot end with $g'(j)=-1$.
```

We initialize both $j_0$ and $j_{-1}$ to $0$, so that they produce no accumulation initially, and that all possibilities of $(0, j_0]$ will be counted if a $-1$ has not occurred yet.

Now, we generalize this onto two dimensions, with an $N\times N$ pasture of $G(i, j)$.
First, we consider the cases where the rectangle starts at $i=1$ and ends at $i=N$.
Selection-wise, this is essentially our $1$-dimensional case.
However, this time, we mark $g'(j)$ as $-1$ if any $i$ has $G(i, j)<100$, as any $-1$ blocks selection.
$g'(j)$ can be marked as $0$ if any $i$ has $G(i, j)=100$, as only $1$ $100$ is needed to select.
Thus, $g'(j)$ will only be $1$ if all $i$'s have $G(i, j)>100$.

We can now generalize to the case where the rectangle can start at any $i=i_1$ and end at any $i=i_2\geq i_1$.

> We loop over $i_1$ in the standard fashion;

Reviewer: "standard fashion" ?

Author: ```suggestion
We loop over all possible $i_1$;
```

at each $i_1$, we reset $g'(j)$ to $1$ (as having no possible $i$'s satisfies "all $i$'s have $G(i, j)>100$).
Then, we loop over $i_2$ and $j$.

> At each $j$, we update $g'(j)$ appropriately to see if $-1$ or $0$ should be marked here due to $G(i,j)$'s value.

Reviewer: ```suggestion
At each $j$, we update $g'(j)$ explicitly to see if $-1$ or $0$ should be marked here due to $G(i,j)$'s value.
```

We must find a range of $j$'s in which there are no $-1$'s, and at least one $0$.
We precompute $g'(j)$.
The objective is now counting the number of intervals with no $-1$'s and at least one $0$.

Next, we sweep $j$ from $1$ to $N$, and keep track of the last index at which we encounter $0$ and $-1$'s as $j_0$ and $j_{-1}$, respectively.
For an interval to be picked, it must include a $0$, so the starting index must be no later than the last $0$ we encounter;
the interval cannot include a $-1$, so the starting index must be later than the last $-1$ we encounter.
Thus, any starting index in the range $(j_{-1}, j_0]$ works (provided $j_{-1}<j_0$), so we simply accumulate our current value of $\max(0, j_0-j_{-1})$ onto the answer if $g'(j)\in{0,1}$ (as we cannot end with $g'(j)=-1$) while we sweep.
We initialize both $j_0$ and $j_{-1}$ to $0$, so that they produce no accumulation initially, and that all possibilities of $(0, j_0]$ will be counted if a $-1$ has not occurred yet.

Now, we generalize this onto two dimensions, with an $N\times N$ pasture of $G(i, j)$.
First, we consider the cases where the rectangle starts at $i=1$ and ends at $i=N$ vertically.

> This is essentially our $1$-dimensional case in terms of selection possibility.

Reviewer 2: I think you shd elaborate on this a bit more something like: Once the top and bottom rows are fixed, the only remaining choice is the left and right columns, so the problem reduces to counting valid intervals in the compressed array.

Author: ```suggestion
Once the top and bottom rows are fixed, the only remaining choice is the left and right columns, so the problem reduces to counting valid intervals horizontally.
```

For $g'$, however, $g'(j)$ should be marked as $-1$ if any $i$ has $G(i, j)<100$, as any $-1$ blocks selection.
$g'(j)$ can be marked as $0$ if any $i$ has $G(i, j)=100$, as only one $100$ is needed to select a range.
This leaves $g'(j)$ as $1$ only if all $i$'s have $G(i, j)>100$.

We now generalize to the case where the rectangle can start at any $i=i_1$ and end at any $i=i_2\geq i_1$.
At each possible $i_1$ we enumerate, we reset $g'(j)$ to $1$ (as having no possible $i$'s satisfies "all $i$'s have $G(i, j)>100$"), before we loop over $i_2$ and $j$.

> At each $j$ in the inner loop, we update $g'(j)$ explicitly to account for $G(i_2,j)$'s, the latest addition to values at column $j$ in our selections.

Reviewer 2: Since g'(j) already summarizes rows [i1..i2-1], when extending the rectangle to row i2 we only need to incorporate the newly added cell G(i2,j).

Author: ```suggestion
Since $g'(j)$ already accounts for $G(i_1\ldots i_2-1,j)$, we update it to explicitly to account for $G(i_2,j)$, the newly added cell, at each $j$ in the inner loop.
```

> After the $g'(j)$ update, we proceed as we would in the $1$-dimensional case.

Reviewer 2: Overall id word this better since the naive way to do this would be to run a check for each column from i1 to i2 but the way that you use uses a g prime array so id briefly go over how each time we add a row we only must run o(n) which optimizes our sol. Also line 35 is p vague imo so add a lil over there but apart from that its a rlly nice edi I liked ur first paragraph cuz it gave the intuition on the sol rlly well so good work!

Author: Thanks!

---
id: usaco-1112
source: USACO Silver 2021 February
title: Just Green Enough
author: John Zhou
---

[Official Analysis (C++, Java)](https://usaco.org/current/data/sol_prob3_silver_feb21.html)

## Explanation

Firstly, let's consider solving this problem in 1D, where we pick intervals from the list $g(j)$, the values for an $1\times N$ pasture.

Let $g'(j)$ be $-1$ if $g(j)<100$, $0$ if $g(j)=100$, and $1$ if $g(j)>100$.
We must find a range of $j$'s in which there are no $-1$'s, and at least one $0$.
We precompute $g'(j)$.
The objective is now counting the number of intervals with no $-1$'s and at least one $0$.

Next, we sweep $j$ from $1$ to $N$, and keep track of the last index at which we encounter $0$ and $-1$'s as $j_0$ and $j_{-1}$, respectively.
For an interval to be picked, it must include a $0$, so the starting index must be no later than the last $0$ we encounter;
the interval cannot include a $-1$, so the starting index must be later than the last $-1$ we encounter.

> Thus, any starting index in the range $(j_{-1}, j_0]$ works (provided $j_{-1}<j_0$), so we simply accumulate our current value of $\max(0, j_0-j_{-1})$ onto the answer if $g'(j)\in{0,1}$ (as we cannot end with $g'(j)=-1$) while we sweep.

Reviewer 3: Here, make sure you are specifying that you are counting intervals that start between j_0 and j_1 and ending at j, and that is why we accumulate simply j_0 - j_(-1).
Currently, you just says sweeping, tracking last indices, and jump to adding j_0 - j_(-1).

Author: Done

We initialize both $j_0$ and $j_{-1}$ to $0$, so that they produce no accumulation initially, and that all possibilities of $(0, j_0]$ will be counted if a $-1$ has not occurred yet.

> Now, we generalize this onto two dimensions, with an $N\times N$ pasture of $G(i, j)$.

Reviewer 3: This sentence is unclear. 
What do you mean by N x N pasture of G(i, j). 

i rows, j colums? going from (1,1) to (i, j)?

Furthemore, G(i, j) is never defined in earlier paragraphs.

Reviewer 2: He means that theres a pasture of size N*N where G(i,j) is the value at cell (i,j) so i mean he did define it but ig its weird wording

Author: It’s defined in prooblem.  I clarified it

First, we consider the cases where the rectangle starts at $i=1$ and ends at $i=N$ vertically.

> Once the top and bottom rows are fixed, the only remaining choice is the left and right columns, so the problem reduces to counting valid intervals horizontally.

Reviewer 3: Again, kinda unclear, there is better phrasing that could be used to merge the two sentences. Maybe something like:

"First, we consider the case where the top and bottom rows of our subrectangle is $i = 1 and i = n$. Then, we are left with choosing left and right columns such that we form a valid subrectangle." (Something like that idk)

Also, it is unclear what "the only remaining choice is the left and right columns". Remaining choice of what?? choosing a subrectangle with top and bottom rows (i, j). Make sure it is clear what ur referencing to, in this case constructing a subrectangle with top/bottom rows i = 1, n.

Sorry If I have misunderstood what you were meaning to say, I'm happy for you to correct me :).

Author: Clarified now

> For $g'$, however, $g'(j)$ should be marked as $-1$ if any $i$ has $G(i, j)<100$, as any $-1$ blocks selection.

Reviewer 3: Use different variables in your 2D section as to not confuse to your 1D section

Author: It serves the same purpose, so I don’t see why.

> For $g'$, however, $g'(j)$ should be marked as $-1$ if any $i$ has $G(i, j)<100$, as any $-1$ blocks selection.

Reviewer 3: Lowkey confusing what you mean, you are mixing a lot of different variables together here, and maybe its also bc of unclarity in previous sections.

To my understanding:
g'(j) -> There is one of these for every row as you go  horizontally in the columns? (1 X N).
G(i, j) is G[i][j], if so i think its less confusing to just say G[i][j].

The section where you say "as any $-1$ blocks selection" is repetitive but ok.

Author: Rewrote this explicity

$g'(j)$ can be marked as $0$ if any $i$ has $G(i, j)=100$, as only one $100$ is needed to select a range.

> This leaves $g'(j)$ as $1$ only if all $i$'s have $G(i, j)>100$.

Reviewer 3: Cant all these three lines just be simplified to
g'(j) = -1 if G[i][j] < 100, g'(j) = 0 if G[i][j] = 100, g'(j) = 1 if G[i][j] > 100. (With proper english of course, but no need for like three seperate lines re-explaining your assigning logic you had in the 1D case.)

Author: Yep, just emphasized the ANY part.

> We now generalize to the case where the rectangle can start at any $i=i_1$ and end at any $i=i_2\geq i_1$.

Reviewer 3: subrectangle/subgrid

Author: Done

> At each possible $i_1$ we enumerate, we reset $g'(j)$ to $1$ (as having no possible $i$'s satisfies "all $i$'s have $G(i, j)>100$"), before we loop over $i_2$ and $j$.

Reviewer 3: Why is "we enumerate" needed?

Author: Removed

Let $g'(j)$ be $-1$ if $g(j)<100$, $0$ if $g(j)=100$, and $1$ if $g(j)>100$.
We must find a range of $j$'s in which there are no $-1$'s, and at least one $0$.
We precompute $g'(j)$.
The objective is now counting the number of intervals with no $-1$'s and at least one $0$.

Next, we sweep $j$ from $1$ to $N$, and keep track of the last index at which we encounter $0$ and $-1$'s as $j_0$ and $j_{-1}$, respectively.
For an interval to be picked, it must include a $0$, so the starting index must be no later than the last $0$ we encounter;
the interval cannot include a $-1$, so the starting index must be later than the last $-1$ we encounter.
Thus, any range with a starting index in the range $(j_{-1}, j_0]$ and an ending index at $j$ works, provided $j_{-1}<j_0$ and $g'(j)\in\{0,1\}$.
Therr are are $j_0-j_{-1}$ intervals; we accumulate those numbers if the conditions are satisfied to get our answer.
We initialize both $j_0$ and $j_{-1}$ to $0$, so that they produce no accumulation initially, and that all possibilities of $(0, j_0]$ will be counted if a $-1$ has not occurred yet.

> Now, we generalize this onto two dimensions, where we pick subrectangles from an $N\times N$ pasture with values $G(i, j)$.

Reviewer 3: Again, can you maybe point out the line or tell me what this sentence means??

Pick subrectangles with $N\times N$ pasture with values $G(i, j)$ to me could have two meanings: subrectangle with the bottom right corner at $(i, j)$ or subrectangles, which are going to have a bunch of values which we can just denote as G(i, j).

Author: ```suggestion
Now, we generalize this onto two dimensions, where we pick subrectangles from an $N\times N$ pasture. We denote the values in the pasture $G(i, j)$ for $i,j\in[1,N]$.
```

First, we consider the cases where the rectangle's bounds include $i=1$ and  $i=N$, leaving us the choice of the interval for the left and right bounds.
As in the 1D case, we sweep from left to right for the right endpoint, and build a $g'$ to track feasibility of the left endpoint by column.

> Any $<100$ element in a column blocks selection, and any $=100& element allows selection, so $g'$ is marked $-1$ or $0$ if these conditions hold for any element in a column, and $1$ otherwise.

Reviewer 3: wrote a & instead of a $

Author: ```suggestion
Any $<100$ element in a column blocks selection, and any $=100$ element allows selection, so $g'$ is marked $-1$ or $0$ if these conditions hold for any element in a column, and $1$ otherwise.
```

> We then proceed as in the 1D case frlm $g'$.

Reviewer 3: what does frlm mean?

Author: ```suggestion
We then proceed as in the 1D case for $g'$.
```
